### PR TITLE
State script

### PR DESCRIPTION
**Pull request by MzScorpio**

### CODE REVIEW

**Comment by d-o-g** [comment]
packages should not repeat, script.state and script.task is sufficient

dont think this should be in the api to begin with
while this is an attempt to abstract state scripts, there's not really a need because state scripts are already usually basic and only a realistic option for smaller scripts. an actual fsm _might_ be better but still, anything beyond that becomes hard to follow and at that point task or any tree based framework becomes superior

i also dont see anyone taking the state script approach except for people who are new to scripting in general, in which case its probably easier for them to approach state scripts the traditional way

regardless, one of the things discussed by devs is the removal of state scripts entirely in the future (possibly even the loop model)

**Comment by MzScorpio** [comment]
> 
> 
> packages should not repeat, script.state and script.task is sufficient

Fixed

> dont think this should be in the api to begin with
> while this is an attempt to abstract state scripts, there's not really a need because state scripts are already usually basic and only a realistic option for smaller scripts. an actual fsm _might_ be better but still, anything beyond that becomes hard to follow and at that point task or any tree based framework becomes superior
> 
> i also dont see anyone taking the state script approach except for people who are new to scripting in general, in which case its probably easier for them to approach state scripts the traditional way

Indeed, state scripts are far more easier to control/debug and due to that I actually find it more useful in big scripts where a task conditions are more likely to overlap.

> regardless, one of the things discussed by devs is the removal of state scripts entirely in the future (possibly even the loop model)

Mmmmmmm

**Comment by MzScorpio** [comment]
It's also good to have in the API since, as you said, most people who are new to scripting use the state approach but they also, most of the tame, have no experience with the object-oriented approach. If we were to agree that task scripts are overall better, we could also agree that having a state script implementation in the API makes it easier for these people to transition to task scripts.

**Comment by buracc** [comment]
People shouldn't write scripts if they can't implement object orientation in their scripts

**Comment by MzScorpio** [comment]
That's not the main point but it's still good for learning.

**Comment by d-o-g** [comment]
if they are new they wont understand object oriented approach to begin with and its arguable that transitioning from this to task script is easier, i think taskscript is more straightforward to begin with (moving on from the loop and if else/switch state) 

even i find this approach confusing to think about logically, perhaps an example can clarify

**Comment by MzScorpio** [comment]
I find it less brain-consuming to decide from within the current state if we should execute something else (another state) or keep doing what we're doing now instead of going over the validation condition for each task and make sure that there's no clash. It gets tedious with big scripts when the wrong task executes. Furthermore, it forces you to remember every single validation condition for all the current tasks when you're adding a new task (Let alone adding that new task a few weeks later since the last time you looked at that script)

As for the mix approach, assume the following for a PK script
- State 1: Fight
- State 2: Run out of wilderness

In both states we will need to check if the player needs to eat which can be done by:
1- Adding the required code to both states
2- Adding a "task" that always does that before executing the state itself

**Comment by Jasper-ketelaar** [comment]
I'm personally not a fan of this being standard in the API either for the following reasons:

1) You will break any script that is compiled using the framework on the SDN when you release a new version where you alter the structure.

2) A lot of the times script require customization of the framework in very specific manners and this can become way more difficult if the framework is integrated instead of provided externally

Personally I'm not even a fan of having the task script be in the bot by default because of the above reasons but I feel like it's a decent stepping stone for people who are getting into script development and want to up the complexity of their scripts slightly before moving on to their own framework.

I think this state script would be great for a nondefault provided API such as the external API we currently have for RSPeer, so it's not like your work goes unappreciated.

**Comment by Lezor91** [comment]
Yeah I agree, even TaskScript will be removed and it was only a thing because we had our own tests/scripts in the repo at some point earlier on
